### PR TITLE
save/restore log level across daemon reexec

### PR DESCRIPTION
**Pull request by poettering**
A rework of @fbuihuu's #9135. It's mostly his patches, however heavily updated. PTAL.

Replaces #9135

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -4448,6 +4469,60 @@ void manager_unref_console(Manager *m) {
...
+                m->original_log_target = log_get_target();
+                m->log_target_overridden = true;
+        }
+
+        log_info("Setting log target to %s.", log_target_to_string(target));
+        log_set_target(target);
+}
+
+void manager_restore_original_log_target(Manager *m) {
+        _cleanup_free_ char *s = NULL;
```
Comment (yuwata) [inline src/core/manager.c]: `s` is not used.
```
../src/core/manager.c:4514:30: warning: unused variable 's' [-Wunused-variable]
        _cleanup_free_ char *s = NULL;
```

Comment (poettering) [reply src/core/manager.c]: indeed!

**Comment by poettering** [comment]
I force pushed a new version now, with the issue @yuwata pointed out fixed. PTAL!

File: man/systemd.xml [hunk 2]
```diff
@@ -749,15 +749,16 @@
...
         <term><constant>SIGRTMIN+23</constant></term>
 
-        <listitem><para>Sets the log level to <literal>debug</literal>
-        (or <literal>info</literal> on
-        <constant>SIGRTMIN+23</constant>), as controlled via
-        <varname>systemd.log_level=debug</varname> (or
-        <varname>systemd.log_level=info</varname> on
-        <constant>SIGRTMIN+23</constant>) on the kernel command
-        line.</para></listitem>
+        <listitem><para>Restores the log level to its default value as defined by the default
```
Comment (keszybz) [inline man/systemd.xml]: One "default" too many in this sentence. Maybe "`Restores the log level to its configured value. That will the in order of priority, the value specified with <varname>systemd.log-level=</varname> on the kernel command line, or the value specified with <option>LogLevel=</option> in the config file, or the built-in default of <literal>info</literal>.`"

File: man/systemd.xml [hunk 3]
```diff
@@ -770,20 +770,19 @@
 
       <varlistentry>
         <term><constant>SIGRTMIN+26</constant></term>
+
+        <listitem><para>Restores the log target to its default value as defined by the default
+        settings.</para></listitem>
```
Comment (keszybz) [inline man/systemd.xml]: Same comment as on the other patch: "default ... default".

File: man/systemd.xml [hunk 4]
```diff
@@ -770,20 +770,19 @@
...
-        <varname>systemd.log_target=journal-or-kmsg</varname> (or
-        <varname>systemd.log_target=console</varname> on
-        <constant>SIGRTMIN+27</constant> or
-        <varname>systemd.log_target=kmsg</varname> on
-        <constant>SIGRTMIN+28</constant>) on the kernel command
-        line.</para></listitem>
+        <listitem><para>Sets the log target to <literal>console</literal> on <constant>SIGRTMIN+27</constant> (or
+        <literal>kmsg</literal> on <constant>SIGRTMIN+28</constant>), as controlled via
+        <varname>systemd.log_target=console</varname> (or <varname>systemd.log_target=kmsg</varname> on
+        <constant>SIGRTMIN+28</constant>) on the kernel command line.</para></listitem>
```
Comment (keszybz) [inline man/systemd.xml]: Hmm, can we change "as controlled via" to "as with"? "as controlled" suggests that this somehow copies what was set with "systemd.log-target=". At least I had to read the sentence a few times to understand.

**Comment by fbuihuu** [comment]
This doesn't work AFAICS:

```
$ systemd-analyze log-level
info
$ mkdir -p /etc/systemd/system.conf.d
$ echo -e '[Manager]\nLogLevel=warning' >/etc/systemd/system.conf.d/log.conf
$ systemctl daemon-reload
$ systemd-analyze log-level
warning
$ systemd-analyze log-level debug
$ systemd-analyze log-level
debug
$ systemctl daemon-reload
$ systemd-analyze log-level
warning
```
As already discussed, we need to take care of the reload path as well.

**Comment by poettering** [comment]
Force pushed a new version with all issues pointed out by @fbuihuu and @keszybz addressed. PTAL!

**Comment by fbuihuu** [comment]
Still doesn't work:

```
$ systemd-analyze log-level debug
$ echo -e '[Manager]\nLogLevel=warning' >/etc/systemd/system.conf.d/log.conf
$ systemctl daemon-reload
$ systemd-analyze log-level
debug
$ systemd-analyze log-level ""
$ systemd-analyze log-level
info
```
It should have returned "warning".

**Comment by fbuihuu** [comment]
arf... @keszybz has just merged this...

This makes me wonder what's the point to request review from my side if they're ignored :-/

**Comment by fbuihuu** [comment]
Also I would have suggested to remove my ownership on these commits since the logic has been rewritten by @poettering.

**Comment by keszybz** [comment]
@fbuihuu sorry, I tested using your reproducer (https://github.com/systemd/systemd/pull/9263#issuecomment-396937079) and it all worked fine here. Let me test again.

**Comment by keszybz** [comment]
```console
[root@krowka ~]# systemd-analyze log-level debug
[root@krowka ~]# echo -e '[Manager]\nLogLevel=warning' >/etc/systemd/system.conf.d/log.conf
[root@krowka ~]# systemctl daemon-reload
[root@krowka ~]# systemd-analyze log-level
debug
[root@krowka ~]# systemd-analyze log-level ""
[root@krowka ~]# systemd-analyze log-level
warning
```
Interesting.

**Comment by fbuihuu** [comment]
@keszybz, make sure the drop in is not already installed before starting the test.

**Comment by keszybz** [comment]
> make sure the drop in is not already installed before starting the test

It wasn't, I removed it after previous round of testing.

Do you get this after `systemd-analyze log-level ""`:
> systemd[1]: Restoring log level to original (warning).

?

I see the following problem:
```console
[root@krowka ~]# systemd-analyze log-level ""
[root@krowka ~]# systemd-analyze log-level
warning
[root@krowka ~]# rm /etc/systemd/system.conf.d/log.conf
rm: remove regular file '/etc/systemd/system.conf.d/log.conf'? y
[root@krowka ~]# systemctl daemon-reload
[root@krowka ~]# systemd-analyze log-level
warning
```
It should be "info", because that's the compiled-in default.

**Comment by fbuihuu** [comment]
> I see the following problem:
> ...
> It should be "info", because that's the compiled-in default.

I think it's a known issue (it's been added recently by @poettering in the TODO list): the compiled-in defaults are not restored when PID is reloading/re-executing.

**Comment by fbuihuu** [comment]
@keszybz I can see:

`systemd[1]: Restoring log level to original (info).`

BTW I did the test with the patches backported on v238. I don't think this should make any difference though.

**Comment by fbuihuu** [comment]
@keszybz, to make sure , could you try to remove the drop-in and reboot before testing again ?

**Comment by poettering** [comment]
i reproduced the issue here. and i agree with @fbuihuu this is caused by use not resetting the manager settings to their built-in defaults before reloading the configuration files.

Or to say this differently: our parse_config_file() code should probably be fixed to only parse the unit-related options, so that "daemon-reload" consistently has only an effect on units, but not on the manager itself.

PID 1 merges configuration data from:

1. it's built-in defaults
2. the configuration file
3. the kernel command line
4. the process command line
5. environment variables

This overlaying of configuration data is messy, and the current code really doesn't do this properly as during reexec we overlay all these, but in case of daemon reload we ignore 1, 3, 4, 5 and only do 2.

But given that this has been like this for a while I think we should leave it like this for v239, and fix this properly later.

**Comment by poettering** [comment]
(and this is further complicated because the env var stuff is partially parsed by main.c and partially by log.c...)

**Comment by fbuihuu** [comment]
@poettering there's still the test case described in https://github.com/systemd/systemd/pull/9263#issuecomment-397206306 which fails here (patches tested on v238 but I don't think this matters).

@keszybz got a different result so it would be interesting which behavior you get.

**Pull request merged**